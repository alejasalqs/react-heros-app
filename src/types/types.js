// Este objeto va a contener todos los types de la aplicacion
// son pares de valores

export const types = {
  login: "[auth] login",
  logout: "[auth] logout",
};
