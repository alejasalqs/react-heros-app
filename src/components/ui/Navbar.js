import React, { useContext } from "react";
import { Link, NavLink, useHistory } from "react-router-dom";
import { AuthContext } from "../../auth/AuthContext";
import { types } from "../../types/types";

// Navbar no es una ruta, es solo un compoenent por lo que no tiene accesos a el prop de history
export const Navbar = () => {
  const { user, dispatch } = useContext(AuthContext);

  // Hook de reac-router para rurilizar el history no por props
  const history = useHistory();

  const handleLogOut = () => {
    dispatch({
      type: types.logout,
    });
    history.replace("/login");
  };
  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <Link className="navbar-brand" to="/">
        Asociaciones
      </Link>

      <div className="navbar-collapse">
        <div className="navbar-nav">
          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/dc"
          >
            DC
          </NavLink>

          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/search"
          >
            Search
          </NavLink>
        </div>
      </div>

      <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
        <ul className="navbar-nav ml-auto">
          <span className="nav-item nav-link">{user.name}</span>
          <button className="nav-item nav-link btn" onClick={handleLogOut}>
            Logout
          </button>
        </ul>
      </div>
    </nav>
  );
};
