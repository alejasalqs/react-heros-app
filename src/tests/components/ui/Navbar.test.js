import React from "react";
import { shallow } from "enzyme";
import { Navbar } from "../../../components/ui/Navbar";
import { AuthContext } from "../../../auth/AuthContext";
import { MemoryRouter, Router } from "react-router";
import { types } from "../../../types/types";

describe("Pruebas en el componente <Navbar/>", () => {
  const historyMock = {
    push: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
    replace: jest.fn(),
  };
  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: "AlejoTest",
    },
  };
  const wrapper = shallow(
    // Es mejor usar mount()
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter>
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </MemoryRouter>
    </AuthContext.Provider>
  ); // Aqui usamos el componente router porque le pasamos el historyMock que creamos para las pruebas

  afterEach(() => {
    // Limpiamos los mocks en cada pruebas
    jest.clearAllMocks();
  });
  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // Esta prueba falla porque no podemos ver dentro del wapper
  test("Debe de mostrar correctamente del nombre de usuario ", () => {
    expect(wrapper.find(".text-info").text().trim()).toBe(
      contextValue.user.name
    );
  });

  test("Debe de llamar el logOut y usar history ", () => {
    // Llamamos y ejecutamos la funcion de click de esta manera
    wrapper.find("button").prop("onClick")();

    expect(contextValue.dispatch).toBeCalledWith({
      type: types.logout,
    });

    expect(historyMock.replace).toHaveBeenLastCalledWith("/login");
  });
});
