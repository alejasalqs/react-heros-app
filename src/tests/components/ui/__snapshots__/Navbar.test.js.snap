// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Navbar/> Debe de mostrarse correctamente 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "createHref": [MockFunction],
        "listen": [MockFunction],
        "location": Object {},
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
  >
    <Navbar />
  </Router>
</MemoryRouter>
`;
